virtual_rainforest/models/soil/carbon.py

```diff
@@ -0,0 +1,203 @@
+
+    def update_pools(
+        self,
+        pH: NDArray[np.float32],
+        bulk_density: NDArray[np.float32],
+        soil_moisture: NDArray[np.float32],
+        soil_temp: NDArray[np.float32],
+        percent_clay: NDArray[np.float32],
+        dt: float,
+    ) -> None:
+        """Update all soil carbon pools.
+
+        This function calls lower level functions which calculate the transfers between
+        pools. When all transfers have been calculated the net transfer is used to
+        update the soil pools.
+
+        Args:
+            pH: pH values for each soil grid cell
+            bulk_density: bulk density values for each soil grid cell
+            soil_moisture: soil moisture for each soil grid cell
+            soil_temp: soil temperature for each soil grid cell
+            percent_clay: Percentage clay for each soil grid cell
+            dt: time step (days)
+        """
+        # TODO - Add interactions which involve the three missing carbon pools
+
+        lmwc_to_maom = self.mineral_association(
+            pH, bulk_density, soil_moisture, soil_temp, percent_clay
+        )
+
+        # Once changes are determined update all pools
+        self.lmwc -= lmwc_to_maom * dt
+        self.maom += lmwc_to_maom * dt
+
+    def mineral_association(
+        self,
+        pH: NDArray[np.float32],
+        bulk_density: NDArray[np.float32],
+        soil_moisture: NDArray[np.float32],
+        soil_temp: NDArray[np.float32],
+        percent_clay: NDArray[np.float32],
+    ) -> NDArray[np.float32]:
+        """Calculates net rate of LMWC association with soil minerals.
+
+        Following Abramoff et al. (2018), mineral adsorption of carbon is controlled by
+        a Langmuir saturation function. At present, binding affinity and Q_max are
+        recalculated on every function called based on pH, bulk density and clay
+        content. Once a decision has been reached as to how fast pH and bulk density
+        will change (if at all), this calculation may need to be moved elsewhere.
+
+        Args:
+            pH: pH values for each soil grid cell
+            bulk_density: bulk density values for each soil grid cell
+            soil_moisture: soil moisture for each soil grid cell
+            soil_temp: soil temperature for each soil grid cell
+            percent_clay: Percentage clay for each soil grid cell
+
+        Returns:
+            lmwc_to_maom: The net flux from LMWC to MAOM
+        """
+
+        # This expression is drawn from (Mayes et al. (2012))
+        binding_affinity = 10.0 ** (
+            BINDING_WITH_PH["slope"] * pH + BINDING_WITH_PH["intercept"]
+        )
+
+        # This expression is also drawn from Mayes et al. (2012)
+        # Original paper also depends on Fe concentration, but we are ignoring this for
+        # now
+        Q_max = bulk_density * 10 ** (
+            MAX_SORPTION_WITH_CLAY["slope"] * np.log10(percent_clay)
+            + MAX_SORPTION_WITH_CLAY["intercept"]
```

Comment:
You could make use of matrix multiplication here to speed this up? I don't know if makes a massive difference with `a + bx` , so maybe not here but this could be:

```python
msc = np.array([2.328, 0.483])

np.sum(
    msc * np.column_stack([np.ones_like(percent_clay), np.log10(percent_clay)]), axis=1
)
```

Reply:
I think it'll probably make things slower tbh, but either way there's not likely to be much in it time-wise and I'd be inclined to leave it as is for clarity's sake.

Reply:
I almost didn't post it 😄 I'd just been reviewing some code where there are several consecutive sums of powers and there it was definitely easier than writing each term!

https://github.com/davidorme/pyrealm/blob/master/pyrealm/pmodel.py#L73-L85